### PR TITLE
[Push] SyncManager and new config options

### PR DESCRIPTION
**Pull request by EmilianoSanchez**
# JS SDK

## What did you accomplish?

- `SyncManager`
- Config options: `streamingEnabled`, `authRetryBackoffBase`, `streamingReconnectBackoffBase`, new urls.
- Update `SplitFactory`
- Other fixes:
  - Update GA type declaration comments.
  - Bug fix: destroying the main client must clean the periodic fetch (MySegments) of all clients. destroying shared clients must only clean their periodic fetch.

## How do we test the changes introduced in this PR?

- [TODO] unit tests

## Extra Notes

### CODE REVIEW

File: src/utils/settings/index.js [hunk 1]
```diff
@@ -139,6 +148,15 @@ function defaults(custom) {
   // `integrations` returns an array of valid integration items.
   withDefaults.integrations = integrations(withDefaults);
 
+  // validate push options
+  if (typeof withDefaults.streamingEnabled !== 'boolean') withDefaults.streamingEnabled = false;
```
Comment (NicoZelaya) [inline src/utils/settings/index.js:152]: Remember you have utilities to use, you should aim to use them as much as possible. Having a single definition of what something is (in this case) allows for easier and faster changes whenever we adapt our rules or want to apply a quick performance optimization.

File: src/utils/settings/index.js [hunk 2]
```diff
@@ -139,6 +148,15 @@ function defaults(custom) {
...
+  if (withDefaults.streamingEnabled === true) {
```
Comment (NicoZelaya) [inline src/utils/settings/index.js:153]: By this line it's a boolean already (if it wasn't, you assigned false on the previous line). Just go with `if (withDefaults.streamingEnabled)`

File: src/utils/settings/index.js [hunk 3]
```diff
@@ -90,6 +90,15 @@ const base = {
...
   integrations: undefined,
+
+  // toggle using (true) or not using (false) Server-Side Events for synchronizing storage
+  streamingEnabled: false,
+
+  // backoff base seconds to wait before re attempting to authenticate for push notifications
+  authRetryBackoffBase: 1,
+
+  // backoff base seconds to wait before re attempting to connect to streaming
+  streamingReconnectBackoffBase: 1,
```
Comment (NicoZelaya) [inline src/utils/settings/index.js:101]: trailing comma.

File: src/sync/node.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
+import PushManagerFactory from './PushManager';
+import { PUSH_DISCONNECT, PUSH_CONNECT } from './constants';
+import FullProducerFactory from '../producer';
+import logFactory from '../utils/logger';
+const log = logFactory('splitio-sync:sync-manager');
```
Comment (NicoZelaya) [inline src/sync/node.js:5]: Might be too manija but try ordering the top section by following a criteria and repeating across all files (not just the ones that mirror each other). For example you could do:
Factories
utilities
CONSTANTS
variables/instances

If you are trying to group by semantics, there is either something wrong with the placement (if the constants are shared, don't put them between factories) 
or with the placement of that constants file (as in it's not for the whole sync but only PUSH hence you have them close, move to the push folder maybe so it's easy to extract as a unit, consider re exporting through PUSH manager if that's the case). 
Consider using spaces or comments if grouping by semantic/functionality, but would be better to stay away of the categories might not always be clear.

Apply this comment to all files here and in other PRs, not necessarily right now (but it's harmless and better to get it off the way) but ideally before closing development.

File: src/sync/node.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  }
+
+  function syncAll() {
+    // fetch splits and segments. There is no need to catch this promise (it is handled by `SplitChangesUpdater`)
+    producer.synchronizeSplits().then(() => {
+      producer.synchronizeSegment();
+    });
+  }
+
+  let pushManager;
```
Comment (NicoZelaya) [inline src/sync/node.js:38]: If you could reorder variables and function declarations too, that's better.

File: src/sync/node.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+  const settings = context.get(context.constants.SETTINGS);
+  if (settings.streamingEnabled)
```
Comment (NicoZelaya) [inline src/sync/node.js:41]: Consider marking the block ({}) if you're adding the line jumps next to the ifs everywhere. Or if you want use them as one liners.

File: src/sync/browser.js [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+ */
+export default function BrowserSyncManagerFactory(mainContext) {
+
+  const contexts = {};
+  let pushManager;
+
+  function startPolling() {
+    log.info('PUSH down or disconnected. Starting periodic fetch of data.');
+    forOwn(contexts, function (context) {
+      const producer = context.get(context.constants.PRODUCER, true);
```
Comment (NicoZelaya) [inline src/sync/browser.js:23]: Do you expect to call this without an actual producer on the context or just being cautious? When do we expect it not to have a producer? 
Can't we use the context returned promise so we kick off polling as soon as the producer is assigned? Or what would happen if you called for start polling but you had no producer and don't schedule using the promise?



Question applies to all gets here, so if one is different, I'd be interesting in knowing it as well.

File: src/sync/browser.js [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    // fetch splits and segments
+    const mainProducer = mainContext.get(mainContext.constants.PRODUCER, true);
+    mainProducer && mainProducer.synchronizeSplits();
+    forOwn(contexts, function (context) {
+      const producer = context.get(context.constants.PRODUCER, true);
+      producer && producer.synchronizeMySegments();
+    });
+  }
+
+  function shared(context, isSharedClient = true) {
```
Comment (NicoZelaya) [inline src/sync/browser.js:50]: This should not be called shared, also the boolean flag is not necessary, specially given the form of the assignment in line #103. 
Invert the parameters, to create the instance you return as main one just call it with (false, context). Then in #103 do (using an example name) `syncManager.shared = createInstance.bind(null, true)` and just call the .shared() method from outside. 

Using a "shared" that creates the full instance and then reexporting as shared property on the instance itself, forcing a boolean telling IT IS a shared instance even though the property accessor is "shared"... seems like entangled.

File: src/sync/browser.js [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+          if (producer && producer.isRunning())
+            producer.stop();
+
+        }
+      }
+    };
+  }
+
+  // called before creating PushManager, to create the producer and put in context.
+  const result = shared(mainContext, false);
```
Comment (NicoZelaya) [inline src/sync/browser.js:96]: "result" is not representative at all. why not just syncManager or mainSyncManager ?

File: src/sync/browser.js [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+      producer && producer.synchronizeMySegments();
+    });
+  }
+
+  function shared(context, isSharedClient = true) {
+    const producer = isSharedClient ? PartialProducerFactory(context) : FullProducerFactory(context);
+    context.put(context.constants.PRODUCER, producer);
+    const settings = context.get(context.constants.SETTINGS);
+    const userKey = matching(settings.core.key);
+    if(contexts[userKey]) log.warn('A client with the same user key has already been created. Only the new instance will be properly synchronized.');
```
Comment (NicoZelaya) [inline src/sync/browser.js:55]: keep your spacing consistent. Both in line jumps and space before the parenthesis. Let's schedule a task later to review the linting and adjust some light autofix rules at some point of Q1.

File: src/sync/browser.js [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+    mainProducer && mainProducer.synchronizeSplits();
+    forOwn(contexts, function (context) {
+      const producer = context.get(context.constants.PRODUCER, true);
+      producer && producer.synchronizeMySegments();
+    });
+  }
+
+  function shared(context, isSharedClient = true) {
+    const producer = isSharedClient ? PartialProducerFactory(context) : FullProducerFactory(context);
+    context.put(context.constants.PRODUCER, producer);
```
Comment (NicoZelaya) [inline src/sync/browser.js:52]: Put the variables towards the top and assignments after unless it makes sense execution~wise or you're defining blocks using spaces/comments. Any of those work.

File: src/utils/context/constants.js [hunk 12]
```diff
@@ -8,7 +8,9 @@ export const SETTINGS = 'settings';
 export const STATUS_MANAGER = 'status_manager';
 export const STORAGE = 'storage';
 export const INTEGRATIONS_MANAGER = 'integrations_manager';
+export const PRODUCER = 'producer';
+export const MY_SEGMENTS_CHANGE_WORKER = 'my_segments_change_worker';
```
Comment (NicoZelaya) [inline src/utils/context/constants.js:12]: These were all in alphabetic order (I missed it when reviewing the integrations one). Please keep them ordered.

File: src/sync/browser.js (hunk 7 above)
Comment (NicoZelaya) [reply src/sync/browser.js:23]: I think we have this pattern in other pieces of the code so let's have a chat about it tomorrow to make sure we're properly in sync about what and specially why I'm asking.

File: types/splitio.d.ts [hunk 13]
```diff
@@ -809,7 +835,27 @@ declare namespace SplitIO {
...
+     * @default 1
+     */
+    authRetryBackoffBase?: number,
+    /**
+     * Seconds to wait before re attempting to connect to streaming.
+     * Next attempts follow intervals in power of two: base seconds, base x 2 seconds, base x 4 seconds, ...
+     * @property {number} streamingReconnectBackoffBase
+     * @default 1
+     */
+    streamingReconnectBackoffBase?: number,
```
[the remark this answers is not in the document]
Comment (EmilianoSanchez) [reply types/splitio.d.ts:858]: the problem is that we don't share these properties with `INodeAsyncSettings`

File: src/sync/node.js (hunk 4 above)
Comment (EmilianoSanchez) [reply src/sync/node.js:5]: thanks. it is good to follow that import order convention

File: src/sync/browser.js (hunk 7 above)
Comment (EmilianoSanchez) [reply src/sync/browser.js:23]: There is always a producer in all contexts when calling `startPolling`, `stopPollingAndSyncAll` and `syncAll`. This is ensure by assigning the producer to the context, before creating the PushManager.
So we can remove the check `if (producer && ...`

Comment (NicoZelaya) [reply src/sync/browser.js:23]: So why do you also pass the true value in the flagCheck param there then?

Comment (EmilianoSanchez) [reply src/sync/browser.js:23]: yes. it is not needed.

**Pull request merged**